File: app/src/core/utility/xmlFormatter.ts
import { XMLBuilder, XMLParser, XMLValidator } from 'fast-xml-parser'
import type { IToolUseCase } from '../ports/IToolUseCase'

export type XmlFormatMode = 'pretty' | 'minify' | 'validate'

export interface XmlFormatterInput {
  source: string
  mode: XmlFormatMode
  indent?: string
}

export interface XmlFormatterResult {
  isValid: boolean
  output?: string
  errorMessage?: string
}

// `preserveOrder` keeps element/text ordering, comments and attributes
// intact through a parse+rebuild round trip -- what a formatter needs,
// unlike the "flatten to a plain object" mode used for data-format
// conversion elsewhere in this module.
const PARSER_OPTIONS = {
  ignoreAttributes: false,
  preserveOrder: true,
  trimValues: true,
}

export class XmlFormatter implements IToolUseCase<XmlFormatterInput, XmlFormatterResult> {
  execute(input: XmlFormatterInput): XmlFormatterResult {
    const validation = XMLValidator.validate(input.source)
    if (validation !== true) {
      return { isValid: false, errorMessage: `Invalid XML: ${validation.err.msg}` }
    }

    switch (input.mode) {
      case 'validate':
        return { isValid: true, output: 'Valid XML' }
      case 'pretty': {
        const output = rebuild(input.source, { format: true, indentBy: input.indent ?? '  ' })
        // The builder emits a leading newline before the root element in
        // pretty mode; strip it so the output starts at the root tag.
        return { isValid: true, output: output.replace(/^\n/, '') }
      }
      case 'minify':
        return { isValid: true, output: rebuild(input.source, { format: false }) }
    }
  }
}

function rebuild(source: string, buildOptions: { format: boolean; indentBy?: string }): string {
  const parser = new XMLParser(PARSER_OPTIONS)
  const parsed = parser.parse(source)
  const builder = new XMLBuilder({ preserveOrder: true, ignoreAttributes: false, ...buildOptions })
  return builder.build(parsed)
}

File: archive/flutter-app/lib/core/utility/xml_formatter.dart
import 'package:xml/xml.dart';

import '../ports/i_tool_use_case.dart';

enum XmlFormatMode { pretty, minify, validate }

class XmlFormatterInput {
  const XmlFormatterInput({
    required this.source,
    required this.mode,
    this.indent = '  ',
  });

  final String source;
  final XmlFormatMode mode;
  final String indent;
}

class XmlFormatterResult {
  const XmlFormatterResult({required this.isValid, this.output, this.errorMessage});

  final bool isValid;
  final String? output;
  final String? errorMessage;
}

class XmlFormatter implements IToolUseCase<XmlFormatterInput, XmlFormatterResult> {
  const XmlFormatter();

  @override
  XmlFormatterResult execute(XmlFormatterInput input) {
    XmlDocument document;
    try {
      document = XmlDocument.parse(input.source);
    } catch (e) {
      return XmlFormatterResult(isValid: false, errorMessage: 'Invalid XML: $e');
    }

    switch (input.mode) {
      case XmlFormatMode.validate:
        return const XmlFormatterResult(isValid: true, output: 'Valid XML');
      case XmlFormatMode.pretty:
        // Collapse whitespace-only text nodes left over from the original
        // formatting so re-indenting doesn't leave stray blank lines.
        document.normalize(trimAllWhitespace: true);
        final output = document.toXmlString(pretty: true, indent: input.indent);
        return XmlFormatterResult(isValid: true, output: output);
      case XmlFormatMode.minify:
        document.normalize(trimAllWhitespace: true);
        final output = document.toXmlString(pretty: false);
        return XmlFormatterResult(isValid: true, output: output);
    }
  }
}
